### PR TITLE
Adding hammer brother **PR: NEED REVIEW**

### PR DESCRIPTION
**Pull request by timmy6336**
# Pull Request Description

Please fill out the form below. For the checkbox, use [x] to have a checkmark. For example:

- [x] With a checkmark.
- [ ] Without a checkmark.

This should not affect the way whether your pull request is accepted or not. Think of it like a survey and documentation, so the reviewers and future developers can follow your PR easier.

1. Have you included *.gitignore* file? Or manually remove config and environment files (E.g: *node_modules* folder, Eclipse IDE files) from this submission? The answer should be **Yes** in this case.

   - [ x] Yes
   - [ ] No

2. Did you fully complete the assigned task(s)?

   - [ ] Fully completed
   - [x ] Partially completed

3. Did you introduce a new non-required feature in this PR? This feature is not an assigned task, but it is the one that you feel will make the game better. For example, reproduce a bug used by speed runner in an original version.

   - [ ] Yes
   - [x ] No

4. Are you aware of any bug caused by your code that have not been fixed in this PR?

   - [ ] Yes
   - [ x] No

5. Are you aware of any **potential bug**? You don't have to fix this, it is just a way to document. You may not face the bug during the development process, but theoretically it can happens in edge cases. For example, memory leak or a player falls out of map.

   - [ ] Yes
   - [x ] No

## Unimplented task(s)

- * some finishing touches are required, the hit box for the hammer needs to be adjusted so it properly follows the head of the hammer and the animation for the hammer brother while throwing needs to be added, the sprite sheet did not have the ful animation and it proved difficult to create it properly from what was provided*

## Related issue(s)

- *Issue number goes here.*
- *If you answer **Yes** to question 3, please provide a description about that in here.*

## Added

- Hammer brother class to enemy.js
- Hammer class to enemy.js

## Changed and Enhanced

- no changes to existing classes
- everything added is independent from existing classes, can just be added to any level

## Fixed

- No fixes to existing classes

## Bugs

1. Not-fixed bugs (only answer when you chose **Yes** in question 4):
   - *Description goes here.*
   - *Description goes here.*
2. Potention bugs (only answer when you chose **Yes** in question 5):
   - *Description goes here.*
   - *Description goes here.*

## Contributors and code author(s)
- *Timmy Roma*

## Final check

Have you fully filled out this document, and add a label **PR: NEED REVIEW** to this pull request? Answer *Yes* (as it should be) means you are good to go! Thanks for the contribution.

- [ x] Yes
- [ ] No

### CODE REVIEW

**Comment by hunghvu** [comment]
@timmy6336 If possible, please consider reformat your PR report and add more information to it. It will make the PRs become more consistent and helpful for future developers/contributors.

File: enemies.js [hunk 1]
```diff
@@ -238,4 +238,322 @@ class PirahnaPlant {
...
+        if(this.game.camera.mario.x < this.x)
+        {
+            this.facing = 1;
+        }
+        else
+        {
+            this.facing = 0;
+        }
+        const FALL_ACC = 1800;
+        //adds in his secret mechanic where after 1 minute he chaces mario
```
Comment (BenjaminDeJager) [inline enemies.js:434]: uh, according to the wiki and my memory, this is NOT part of the game.
https://www.mariowiki.com/Hammer_Bro#Super_Mario_Bros.

feel free to leave the code in, but it should not be on by default which this code seems to be (see this.beserk set to 3600 which is definitely more then 0)

**Review by BenjaminDeJager — CHANGES_REQUESTED** [review]
according to the wiki not mentioning ANYWHERE about a "berserk mode" and my memory, this is NOT part of the game. feel free to leave this code in as its not harmful per-say, but chris has stated that he intends this game to be faithful so please make sure its not on by default.

https://www.mariowiki.com/Hammer_Bro#Super_Mario_Bros.

otherwise it seems fine coding-wise.

**Comment by timmy6336** [comment]
https://www.youtube.com/watch?v=nih0xD0ZtTs

might not have used the exact right timing but it appeared to be about 1 minute

**Comment by hunghvu** [comment]
@timmy6336 Can you show me a way to demo your code?

**Comment by timmy6336** [comment]
when i tested it i added a function loadtest to scenemanger 
 loadTest()
    {
        this.game.addEntity(new Ground(this.game,0,700,1000));
        this.game.addEntity(new Brick(this.game, 50,500,1,"None"));
        this.game.addEntity(new Brick(this.game, 100,500,1,"None"));
        this.game.addEntity(new Brick(this.game, 150,500,1,"None"));
        this.game.addEntity(new Brick(this.game, 200,500,1,"None"));
        this.game.addEntity(new Brick(this.game, 250,500,1,"None"));
        this.game.addEntity(new Brick(this.game, 300,500,1,"None"));
        this.game.addEntity(new Brick(this.game, 350,500,1,"None"));
        this.game.addEntity(new Brick(this.game, 400,500,1,"None"));
        this.game.addEntity(new Brick(this.game, 50,300,1,"None"));
        this.game.addEntity(new Brick(this.game, 100,300,1,"None"));
        this.game.addEntity(new Brick(this.game, 150,300,1,"None"));
        this.game.addEntity(new Brick(this.game, 200,300,1,"None"));
        this.game.addEntity(new Brick(this.game, 250,300,1,"None"));
        this.game.addEntity(new Brick(this.game, 300,300,1,"None"));
        this.game.addEntity(new Brick(this.game, 350,300,1,"None"));
        this.game.addEntity(new Brick(this.game, 400,300,1,"None"));
        this.game.addEntity(new HammerBro(this.game,200,600,1));
        this.game.addEntity(new HammerBro(this.game,300,300,0));
        
        this.mario.x = 0;
        this.mario.y = 0;
        this.game.addEntity(this.mario);
    }

replacing the load level one, or you could just add a hammer brother to a scene like so
this.game.addEntity(new HammerBro(this.game,300,300,0));
or to a json file like so
 hammerBro: [{ x: 50, y: 12.5, facing: 0}]

they are independent so adding them anywhere should cause no issue

File: enemies.js [hunk 2]
```diff
@@ -238,4 +238,322 @@ class PirahnaPlant {
...
+        this.animations.push(new Animator(this.spritesheet, 119, 89, 16, 24, 2, 0.1, 14, false, true));
+        this.paused = true;
+        this.dead = false;
+        this.deadCounter = 0;
+        //this.updateBB();
+        this.swap = 60;
+        this.jump = 240;
+        this.jumpType = 1;
+        this.passThrough = 0;
+        this.throw = 30;
```
Comment (hunghvu) [inline enemies.js:407]: The naming convention here can be improve. When you have a variable with similar names like *throw*, *throws*, *throwing* without a clear comment, this adds noise to your code. Besides, many of your counters are frame based (the amount of update being called). As we cannot consistently get exactly 60 fps, your comments (which use "seconds") can be misleading. From what I know, an original Mario mechanism is heavily frame based, that's why we have terminology like *perfect frame jump*, etc. Thus the implementation here might not exactly be a bad thing.

**Review by hunghvu — DISMISSED** [review]
The CPU seems to be well implemented. However, there are not many interactions at this point. For example, collision with Mario, fire ball, etc. This is acceptable though. I will approve this PR as there is no visible game breaking bug.

**Comment by hunghvu** [comment]
~~Do not merge yet, my merge commit broke the code. It is fixed now but I'm trying to find a way to update this PR.~~

**Review by hunghvu — APPROVED** [review]

**Review by algorithm0r — APPROVED** [review]

**Pull request merged**